### ISSUE TITLE
GAT successfully compiled without error even for unmatched types

### ISSUE THREAD

**Issue by h-a-n-a** [rust-lang/rust#104686]
I tried this code:

```rust
trait PointerFamily {
  type Pointer<T>: Deref<Target = T>;

  fn new<T>(item: T) -> Self::Pointer<T>;
}

struct ArcFamily;

impl PointerFamily for ArcFamily {
  type Pointer<T> = Arc<T>;

  fn new<T>(item: T) -> Self::Pointer<T> {
    Arc::new(item)
  }
}

struct Family<P: PointerFamily<Pointer<String> = Arc<String>>> {
  p: P::Pointer<u32>
}

fn main() {
   let f: Family<ArcFamily> = Family {
      p: ArcFamily::new(123)
   }
}
```

I expected to see this happen: 

The compiler should complain that `struct Family` should not be successfully compiled as type `P::Pointer<u32>` is not matching with `Pointer<String>`

Instead, this happened: 

The snippet was compiled successfully.

Also, it is good to note that if I replace `Pointer<String> = Arc<String>` with `Pointer<String> = Rc<String>`, the compiler failed to compile this snippet as expected. But the compiler seems not doing checks on the generic of the associated type.

### Meta

`rustc --version --verbose`:
```rustc 1.66.0-nightly (4b8f43199 2022-10-19)
binary: rustc
commit-hash: 4b8f4319954ff2642690b9e5cbe4af352d095bf6
commit-date: 2022-10-19
host: aarch64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.2
```

<details><summary>Backtrace</summary>
<p>

```
Compiled successfully
```

</p>
</details>

**Comment by compiler-errors** [comment]
@h-a-n-a -- I don't understand what the bug is here. Can you elaborate?

When you substitute the generic args of `Family<ArcFamily>`, that means that the `p`  field's type is `<ArcFamily as Family>::Pointer::<u32>`, which is `Arc<u32>`.

The only thing that `<P: PointerFamily<Pointer<String> = Arc<String>>>` checks is that when `Pointer<T>` is substituted with `T = String`, we get `Arc<String>` out of it.

**Comment by h-a-n-a** [comment]
> @h-a-n-a -- I don't understand what the bug is here. Can you elaborate?
> 
> When you substitute the generic args of `Family<ArcFamily>`, that means that the `p` field's type is `<ArcFamily as Family>::Pointer::<u32>`, which is `Arc<u32>`. This program doesn't seem to have a type error in it.
> 
> The only thing that `<P: PointerFamily<Pointer<String> = Arc<String>>>` checks is that when `Pointer<T>` is substituted with `T = String`, we get `Arc<String>` out of it.

@compiler-errors 
The confusing part of this, for me, is for the `P: PointerFamily<Pointer<String> = Arc<String>>`, which has an underlying type conflict for the implementing part as `Arc<u32>` was used, and if I tried to write the full type of `p` field, It would be `P::Pointer<u32>`. I don't know if it was an intended behavior, but it looks weird to me though.

**Comment by fmease** [comment]
I can only rephrase what Michael wrote.

The trait bound `P: PointerFamily<Pointer<String> = Arc<String>>` neither states that `P::Pointer` only accepts the type `String` as its argument nor that it “equals” `Arc<String>`. It states that given the concrete type `String` as an argument to `P::Pointer`, `P::Pointer<String>` equals `Arc<String>`. Most notably, we are only talking about `String` here and aren't saying anything about all the other types in existence. E.g. we aren't specifying the shape of `P::Pointer<i32>`. The latter could theoretically be anything.

Imagine `P::Pointer` was a function (conceptually it *is*). You are basically saying struct `Family` takes any “function” `P::Pointer` which happens to returns `Arc<String>` when applied to `String`. This latter fact notwithstanding, you “call” `P::Pointer` just with argument `i32`.

The bound is not very useful. In the future we will hopefully support higher-ranked types and support the bound `P: for<T> PointerFamily<Pointer<T> = Arc<T>>` which quantifies over all types `T` forcing `P::Pointer<T>` to be `Arc<T>` for all types `T`. That'd be way more useful.
